### PR TITLE
BAH-4638| Fix. Tile rendering when there's limited privilege access.

### PR DESCRIPTION
**Pull request by arshiyaTW2021**

### CODE REVIEW

**Review by ravinderkabli — COMMENTED** [review]
## Review for BAH-4638

Good progress on lifting privilege filtering from `AppTile` to the grid level — the architectural direction is sound, and the test coverage is thorough. Found a race condition with the provider's initial state that needs fixing.

Found 1 blocking issue, 2 suggestions, and 2 nits.

**Acceptance Criteria**: 4/4 in-scope ACs met (7, 8, 9, 6), but the race condition (issue #1) can show a spurious error on mount.

### Issues not in diff

| File:Line | Type | Issue |
|-----------|------|-------|
| HomePageGrid.tsx:89 | question | `onClose={() => setError(null)}` dismisses error but doesn't retry loading — user lands on "No modules available" with no recovery path |

File: distro/src/app/components/HomePageGrid/HomePageGrid.tsx [hunk 1]
```diff
@@ -1,23 +1,36 @@
...
+    error: privilegeError,
+  } = useUserPrivilege();
   const [modules, setModules] = useState<Module[]>([]);
-  const [loading, setLoading] = useState(true);
+  const [loading, setLoading] = useState(false);
   const [error, setError] = useState<string | null>(null);
 
   const loadModules = useCallback(async () => {
+    if (privilegesLoading) return;
+    if (privilegeError || userPrivileges === null) {
```
Comment (ravinderkabli) [inline distro/src/app/components/HomePageGrid/HomePageGrid.tsx]: blocking: Race condition — `userPrivileges === null` fires before provider has loaded.

`UserPrivilegeProvider` starts with `{ isLoading: false, userPrivileges: null }` (line 20 of `UserPrivilegeProvider.tsx`). Since child effects fire before parent effects in React, `HomePageGrid`'s `useEffect` sees null privileges with `isLoading=false` on the first cycle and enters this error path before any fetch has started.

Suggested fix — treat null-without-error as "still loading":
```tsx
if (privilegesLoading || (userPrivileges === null && !privilegeError)) return;
if (privilegeError) {
  setError(t('HOME_ERROR_FETCH_CONFIG'));
  return;
}
```

File: distro/src/app/components/HomePageGrid/HomePageGrid.tsx [hunk 2]
```diff
@@ -1,23 +1,36 @@
...
 export const HomePageGrid: React.FC = () => {
   const { t } = useTranslation();
+  const {
+    userPrivileges,
+    isLoading: privilegesLoading,
+    error: privilegeError,
+  } = useUserPrivilege();
   const [modules, setModules] = useState<Module[]>([]);
-  const [loading, setLoading] = useState(true);
+  const [loading, setLoading] = useState(false);
```
Comment (ravinderkabli) [inline distro/src/app/components/HomePageGrid/HomePageGrid.tsx:17]: suggestion: `useState(false)` causes a flash of "No modules available" on mount.

On first render, `privilegesLoading || loading` is `false || false`, so the skeleton is skipped and the component falls through to `modules.length === 0` → empty state. The previous `useState(true)` ensured the skeleton showed until data arrived.

Consider restoring `useState(true)`, or adding `userPrivileges === null` to the loading guard:
```tsx
if (privilegesLoading || loading || userPrivileges === null) {
  // show skeleton
}
```

File: distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx [hunk 3]
```diff
@@ -119,4 +188,107 @@ describe('HomePageGrid', () => {
...
+    render(<HomePageGrid />);
+
+    await waitFor(() => {
+      expect(screen.getByTestId('app-tile-reports')).toBeInTheDocument();
+    });
+
+    expect(screen.getByText('Reports')).toBeInTheDocument();
+  });
+
+  it('shows error when privilege fetch fails', async () => {
```
Comment (ravinderkabli) [inline distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx:233]: suggestion: Missing test for the provider's actual initial state.

`UserPrivilegeProvider` starts with `{ isLoading: false, userPrivileges: null, error: null }`. No test covers this exact combination — the "privilege error" test below uses a non-null error. Adding a test for this state would catch the race condition described above.

File: distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx [hunk 4]
```diff
@@ -119,4 +188,107 @@ describe('HomePageGrid', () => {
...
+    mockUseUserPrivilege.mockReturnValue({
+      userPrivileges: null,
+      setUserPrivileges: jest.fn(),
+      isLoading: false,
+      setIsLoading: jest.fn(),
+      error: 'Privilege fetch failed',
```
Comment (ravinderkabli) [inline distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx]: nit: Type mismatch — `error` should be `Error | null`, not a string.

`UserPrivilegeContextType.error` is typed `Error | null` (see `UserPrivilegeContext.tsx:9`). This should be `error: new Error('Privilege fetch failed')`.

File: distro/src/services/moduleService.ts [hunk 5]
```diff
@@ -53,7 +53,7 @@ export const filterByPrivilege = (
   extensions: Module[],
   userPrivileges?: string[],
 ): Module[] => {
```
Comment (ravinderkabli) [inline distro/src/services/moduleService.ts:55]: nit: The semantic change from `[]` → show all to `[]` → hide privileged is correct but non-obvious.

Consider an inline comment explaining the contract:
```ts
// undefined = skip privilege check (caller hasn't loaded privileges)
// []        = user has zero privileges (show only public tiles)
if (userPrivileges === undefined) {
```

File: distro/src/app/components/HomePageGrid/HomePageGrid.tsx (hunk 1 above)
Comment (arshiyaTW2021) [reply distro/src/app/components/HomePageGrid/HomePageGrid.tsx]: Fixed

File: distro/src/app/components/HomePageGrid/HomePageGrid.tsx (hunk 2 above)
Comment (arshiyaTW2021) [reply distro/src/app/components/HomePageGrid/HomePageGrid.tsx:17]: Done

File: distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx (hunk 3 above)
Comment (arshiyaTW2021) [reply distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx:233]: Done

File: distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx (hunk 4 above)
Comment (arshiyaTW2021) [reply distro/src/app/components/HomePageGrid/__tests__/HomePageGrid.test.tsx]: Done

File: distro/src/services/moduleService.ts (hunk 5 above)
Comment (arshiyaTW2021) [reply distro/src/services/moduleService.ts:55]: Done

**Review by ravinderkabli — APPROVED** [review]
Approved
